presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java

```diff
@@ -76,10 +76,11 @@ public void testNotEliminateSorts()
         PlanMatchPattern pattern =
                 anyTree(
                         sort(
-                                window(windowMatcherBuilder -> windowMatcherBuilder
-                                                .specification(windowSpec)
-                                                .addFunction(functionCall("row_number", Optional.empty(), ImmutableList.of())),
-                                        anyTree(LINEITEM_TABLESCAN_Q))));
+                                anyTree(
```

Comment:
This test uses the following query

`SELECT quantity, row_number() OVER (ORDER BY quantity) FROM lineitem ORDER BY tax`

for which the new plan is

```
presto:tiny> explain SELECT quantity, row_number() OVER (ORDER BY quantity) FROM lineitem ORDER BY tax;
explain SELECT quantity, row_number() OVER (ORDER BY quantity) FROM lineitem ORDER BY tax;
                                                               Query Plan                                                               
----------------------------------------------------------------------------------------------------------------------------------------
 - Output[quantity, _col1] => [quantity:double, row_number_1:bigint]                                                                    
         _col1 := row_number_1                                                                                                          
     - Project[] => [quantity:double, row_number_1:bigint]                                                                              
*         - RemoteMerge[tax ASC_NULLS_LAST] => [quantity:double, tax:double, row_number_1:bigint]                                        
*             - LocalMerge[tax ASC_NULLS_LAST] => [quantity:double, tax:double, row_number_1:bigint]                                     
*                 - PartialSort[tax ASC_NULLS_LAST] => [quantity:double, tax:double, row_number_1:bigint]                                
*                     - RemoteExchange[REPARTITION] => quantity:double, tax:double, row_number_1:bigint                                  
                         - Window[order by (quantity ASC_NULLS_LAST)] => [quantity:double, tax:double, row_number_1:bigint]             
                                 row_number_1 := row_number()                                                                           
                             - LocalExchange[SINGLE] () => quantity:double, tax:double                                                  
                                     Cost: {rows: 60175 (1.15MB), cpu: 1203500.00, memory: 0.00, network: 1203500.00}                   
                                 - RemoteExchange[GATHER] => quantity:double, tax:double                                                
                                         Cost: {rows: 60175 (1.15MB), cpu: 1203500.00, memory: 0.00, network: 1203500.00}               
                                     - TableScan[tpch:tpch:lineitem:sf0.01, originalConstraint = true] => [quantity:double, tax:double] 
                                             Cost: {rows: 60175 (1.15MB), cpu: 1203500.00, memory: 0.00, network: 0.00}                 
                                             quantity := tpch:quantity                                                                  
                                             tax := tpch:tax                                                                                                                                          
```

My interpretation of this plan is that we first gather all the data on a single node, sort it as part of `window` operator, then repartition sorted data, sort locally, merge sort into a single node again. The steps marked with (*) above appear unnecessary. For this type of query, I'd expect the new plan to have a window operator over merge-gather exchange over partial sort. Am I reading this wrong?

Reply:
I think at output you want data to be sorted by `tax`. If you move window operator above final sort then data won't be sorted on `tax` anymore, which doesn't match expected output ordering.